NEP: Add NEP 50 draft about fixing scalar promotion rules (#21103)
numpy/numpy: doc/neps/nep-0050-scalar-promotion.rst

While these choices were well intended, they lead to a complexity both
for users and maintainers.

There are two main ways this can lead to confusing results:

1. This means that the value can matter::

     np.result_type(np.int8, 1) == np.int8
     np.result_type(np.int8, 255) == np.int16

   which is even true when replacing the value with 0-D arrays::

     int64_0d_array = np.array(1, dtype=np.int64)
     np.result_type(np.int8, int64_0d_array) == np.int8
    
   This logic arises, because ``1`` can be represented by an ``int8`` while
   ``255`` can be represented by an ``int16`` *or* ``uint8``.
   This means that for the 0-D array or NumPy scalars, the type is largely ignored.

2. For a Python integer or float the value matters except when the other type is
   also scalar or zero dimensional::

     np.result_type(np.int8(1), 1) == np.int64

   Because in this case value-based promotion is disabled causing the Python 1
   to be considered an int64 (or default integer).

Note that the examples apply just as well for operations like multiplication,
addition, or comparisons and the corresponding functions like `np.multiply`.

> This NEPs proposes to refactor the behaviour around two guiding principles:

Reviewer: NEPs -> NEP I think

1. The value must never influence the result type.
2. NumPy scalars or 0-D arrays must always lead to the same behaviour as
   their N-D counterparts.

> We propose to removes all value-based logic and add special handling for

Reviewer: `removes` -> `remove`

Python scalars to preserve some of the convenience that it provided.

> This changes also apply to ``np.can_cast(100, np.int8)``, however, we expect

Reviewer: `This` -> `These` maybe

   well as making it available to user defined DTypes through
   :ref:`NEP 42 <NEP42>`.
   Currently, this leads to a dual implementation of a new and an old (value
   sensitive) system.  Fixing this will greatly simplify the internal logic
   and make results more consistent.

Usage and Impact
================

There will be no transition period due to the difficulty and noise this is
expected to create.  In rare cases users may need to adjust code to avoid
reduced precision or incorrect results.

We plan to provide an *optional* warning mode capable of notifying users of
potential changes in behavior in most relevant cases.

Impact on ``can_cast``
----------------------

Can cast will never inspect the value anymore.  So that the following results
are expected to change from ``True`` to ``False``::

  np.can_cast(100, np.uint8)
  np.can_cast(np.int64(100), np.uint8)
  np.can_cast(np.array(100, dtype=np.int64), np.uint8)

We expect that the impact of this change will be small compared to that of
the following changes.

.. note::

    The first example where the input is a Python scalar could be preserved
    to some degree, but this is not currently planned.

> Impact on operators functions involving NumPy arrays or scalars

Reviewer: `operators` -> `operator` maybe

Author: The `and` was missing, hopefully that makes more sense then :).  Duplicate the example with `np.add` also.

---------------------------------------------------------------

The main impact on operations not involving Python scalars (float, int, complex)
will be that 0-D arrays and NumPy scalars will never behave value-sensitive.
This removes currently surprising cases.  For example::

  np.arange(10, dtype=np.uint8) + np.int64(1)

Will return an int64 array because the type of ``np.int64(1)`` is strictly
honoured.

Impact on operators involving Python ``int``, ``float``, and ``complex``
------------------------------------------------------------------------

This NEP attempts to preserve most of the convenience that the old behaviour

> gave for Python operators, but remove it for

Reviewer: incomplete thought at the end of this sentence perhaps?

Author: Thanks, mixup there... also a bit of a leftover, because in the first draft, I had angled it for using "weak" logic only on Python operators, which is currently the "alternative".

The current value-based logic has some nice properties when "untyped" Python

> scalars involved::

Reviewer: "are" involved (I think)

  np.arange(10, dtype=np.int8) + 1  # returns an int8 array

>   nparray([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array

Reviewer: `np.array`

But led to complexity when it came to "unrepresentable" values:

  np.arange(10, dtype=np.int8) + 256  # returns int16

>   nparray([1., 2.], dtype=np.float32) * 1e200  # returns float64

Reviewer: `np.array`

The proposal is to preserve this behaviour for the most part.  This is achieved
by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
in these operations.
To mitigate user surprises, we would further make conversion to the new type
more strict.  This means that the results will be unchanged in the first
two examples.  For the second one, the results will be the following::

  np.arange(10, dtype=np.int8) + 256  # raises a TypeError

>   nparray([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity

Reviewer: `np.array`

Reviewer: Not sure how much value there is in it, but for the second changed case that returns infinity, are there examples from other frameworks/programming languages that do something similar in a matching scenario?

I don't know that we'd necessarily want to mimic what other major languages/projects do, though citing them as examples may help make a stronger case if needed. Perhaps in this case one could just quickly verify in C or C++ or whatever.

Author: I added an `admonition`, C normally does this, but of course you can set the evironment to raise an error (just like you can in NumPy, although casts currently don't check that normally!).

I decided to promise to fix that it gives a warning right now.  `pytorch` seems to just overflow silently, I guess they probably don't give warnings reliably, because some hardware they care about doesn't really support it well.

The second one will warn because ``np.float32(1e200)`` overflows to infinity.
It will then do the calculation with ``inf`` as normally.

Impact on functions involving Python ``int``, ``float``, and ``complex``
------------------------------------------------------------------------

Most functions, in particular ``ufuncs`` will also use this weakly typed
logic.
In some cases, functions will call `np.asarray()` on inputs before any operations

> and thus will s

Reviewer: Incomplete thought here for now I think

Author: Sorry, this whole paragraph should have been deleted for now, because I moved the option of handling ufuncs differently to the "alternatives".

.. note::

    There is a real alternative to not do this for `ufuncs` and limit the special
    behaviour to Python operators.  From a user perspective, we assume that most
    functions effectively call `np.asarray()`.
    Because Python operators allow more custom logic, this would ensure that an
    overflow warning is given for all results with decreased precision.

Backward compatibility
======================

In general, code which only uses the default dtypes float64, or int32/int64
or more precise ones should not be affected.

However, the proposed changes will modify results in quite a few cases where
0-D or scalar values (with non-default dtypes) are mixed.
In many cases, these will be bug-fixes, however, there are certain changes
which may be particularly interesting.

The most important failure is probably the following example::

  arr = np.arange(100, dtype=np.uint8)  # storage array with low precision
  value = arr[10]

  # calculation continues with "value" without considering where it came from
  value * 100

Where previously the ``value * 100`` would cause an up-cast to int32/int64
(because value is a scalar).  The new behaviour will preserve the lower

> precision unless explicitly dealt with (just as if ``value`` was an array).

Reviewer: I suppose it would be good to establish that for a number of folks this new behavior would match with the "principle of least surprise." I think the old/current behavior could be convenient for correcting accidents, but perhaps a bit surprising in the context of what one would typically expect for an extracted element of a particular type.

Author: Lets leave this one open.  I like the "principle of least surprise", but I am not sure I want to burden the backcompat section with it, but rather "advertise" it in the motivation section.

This can lead to integer overflows and thus incorrect results beyond precision.
In many cases this may be silent, although NumPy usually gives warnings for the
scalar operators.

Similarliy, if the storage array is float32 a calculation may retain the lower
float32 precision rather than use the default float64.

> Further issues can occure.  For example:

Reviewer: `occure` -> `occur`

* Floating point comparisons, especially equality, may change when mixing
  precisions:
  ```python3
  np.float32(1/3) == 1/3  # was False, will be True.
  ```
* Certain operations are expected to start failing:
  ```python3

>   np.array([1], np.uint8) * 1000

Reviewer 2: Can you clarify, this fails because `1000` is not representable in `uint8`. But am I understanding correctly that `np.array([1], np.uint8) * 100` would silently overflow, because `100` is representable in `uint8`?

My main area of confusion from the set of examples here is which operations will silently overflow and which will error.

Author: Yes, that is what I have in mind currently.  I am not at all set on it though – it would actually be much simpler to not do that!
Basically, my rough plan would be that the above would fail, but `np.uint8(1000)` would not for now, so that you could still write `np.uint8(-1)` if you must for some reason – because... backward compatibility mostly.

That is painful, so a simpler middle ground could also be that both `np.uint8(1000)` and `np.uint8(1) + 1000` would give the same warning.

In general, NumPy does not give reliable warnings for integer overflows.  We only give warnings for scalar python operations (and even then randomly, because scalar operator calls are poorly organized).  So `np.uint8(10) + 250` will give a warning, but that example would not (it works with an array).

That is one argument for why things could be a smoother transition if we limit to Python operators:  For those (and scalars which are currently the place that upcasts a lot), we could pull off to ensure warnings are given.  (Super painful because it requires cleaning up the scalar mess, but a clear path.)

One thing that I am unsure about what JAX does, is that I don't really think we should tag on the "weak" property to the array as JAX does.  One, because I am worried it may be brittle/surprising to carry it around for long and second because unlike JAX our arrays are mutable, so there is an additional complexity.  (We can still tag on something as an implementation detail.)

Reviewer 2: Thanks - regarding JAX & the weak type attribute: the reason we used that approach is that in JAX's JIT-compiled code, everything is represented by a JAX array (scalar inputs get automatically promoted). So to maintain the same behavior between JIT and non-JIT code paths, we need a way to mark a JAX array as weakly-typed. Since NumPy doesn't have that requirement, it seems that it would make sense to avoid it.

That said, one benefit of JAX's approach is that calling `jnp.asarray` on a function's inputs does not change the results even if the input is a scalar, because `asarray` maintains the weakness of the array.

On the other hand, allowing zero-dimensional arrays to be weak readily leads to cases where larger arrays can be weak as well (e.g. `jnp.broadcast_to(jnp.array(1.0), (100, 100))` returns a weakly-typed matrix) so weak types can no longer be thought of as strictly synonymous with scalars. That's a bit surprising, and is one reason we're currently actively discussing design considerations around the semantics of weakly-typed values.

Author: Ah, that is interesting (and I should mention the weak attribute in alternatives, even if I currently would need some convincing to consider it more seriously – and add 1-2 more examples here for clarity of the initial point, of course).

The broadcast example is nicely clear about how things may get confusing when dragging on the flag.  Long ago, I had early plans to support, in theory, even `[1, 2, 3]` to be considered "weak" – e.g. for ufuncs – but I had abandoned them: it seemed just to much and also way too many moving parts at the time.

In the end, it would be nice to have a clear concept of how far the "weakness" reaches.  I could half imagine an `np.asarray(x, allow_weak=True)` for library use, but I somewhat imagine libraries are better of handling it up-front with `np.result_type` (or by not calling `asarray` at all) – possibly `asarray_unless_literal(x)` if need be.

> That's a bit surprising, and is one reason we're currently actively discussing design considerations around the semantics of weakly-typed values.

It will be interesting to hear about that!  NumPy might end up being a bit more limited in options due to backcompat, but it would be nice if we can manage to align some concepts, i.e. at which step "weakness" is forgotten.

Proposed Weak Promotion
-----------------------

This proposal uses a "weak scalar" logic.  This means that Python ``int``, ``float``,
and ``complex`` are not assigned one of the typical dtypes, such as float64 or int64.
Rather, they are assigned a special abstract DType, similar to the "scalar" hierarchy
names: Integral, Floating, ComplexFloatin.

When promotion occurs (as it does for ufuncs if no exact loop matches),
the other DType is able to decide how to regard the Python
scalar.  E.g. a ``UInt16`` promoting with an `Integral` will give ``UInt16``.

.. note::

    A default will most likely be provided in the future for user defined DTypes.
    Most likely this will end up being the default integer/float, but in principle
    more complex schemes could be implemented.

At no time is the value used to decide the result of this promotion.  The value is only
considered when it is converted to the new dtype; this may raise an error.

Related Work
============

* `JAX promotion`_ also uses the weak-scalar concept.  However, it makes use
  of it also for most functions.  JAX further stores the "weak-type" information

>   on the array: ``jnp.array(1)`` remains weakly typed.

Reviewer: Heh, the JAX promotion docs are pretty detailed. I was thinking we might want to briefly mention if our new/proposed direction was consistent with some other major languages/frameworks, though not sure if that changes much.

Author: Yeah, JAX uses the weak logic, I think I should be reporting some bugs to them, but I don't remember (I believe if you swapped the inputs on some of their example, things got wrong, and JAX tags the "was scalar" to the 0-D array, so it can be passed around, which I am not sure always turns out right, but is easier for them because their arrays are immutable.

I believe most use "weak" logic (torch definitely does), although all of these skip scalars entirely probably.

scalar operators.

Similarliy, if the storage array is float32 a calculation may retain the lower
float32 precision rather than use the default float64.

Further issues can occur.  For example:

* Floating point comparisons, especially equality, may change when mixing
  precisions:
  ```python3
  np.float32(1/3) == 1/3  # was False, will be True.
  ```
* Certain operations are expected to start failing:
  ```python3
  np.array([1], np.uint8) * 1000
  np.array([1], np.uint8) == 1000  # possibly also
  ```
  to protect users in cases where previous value-based casting led to an
  upcast.  (Failures occur when converting ``1000`` to a ``uint8``.)
* Floating point overflow may occur in odder cases:
  ```python3
  np.float32(1e-30) * 1e50  # will return ``inf`` and a warning
  ```
  Because ``np.float32(1e50)`` returns ``inf``.  Previously, this would return
  a double precision result even if the ``1e50`` was not a 0-D array

In other cases, increased precision may occur.  For example::

  np.multiple(float32_arr, 2.)
  float32_arr * np.float64(2.)

Will both return a float64 rather than float32.  This improves precision but
slightly changes results and uses double the memory.

> 

Author: Note to self:  A weird corner case is that `arr + 10**100` is currently "obviously" an `object` array result.  We either need to decide to preserve this or note here that users may be forced to write `np.asarray(10**100)` to trigger the slightly awkward "use object if necessary" logic – or provide `dtype=object`.

Reviewer 3: I think automatic promotion to dtype object is always a bad idea. Why would this not just overflow?

Author: Added a subsection to the backwards compatibility part (below), so "resolving" for now

   give fairl equivalent results and 1. could be extended by 2. in the future
   if necessary.

It further requires removing all current special value-based code paths.

Unintuitively, a larger step in the implementation may be to implement a
solution to allow an error to be raised in the following example::

   np.arange(10, dtype=np.uint8) + 1000

Even though ``np.uint8(1000)`` returns the same value as ``np.uint8(232)``.

.. note::

    See alternatives, we may yet decide that this silent overflow is acceptable
    or at least a separate issue.

Alternatives
============

There are several design axes where different choices are possible.
The below sections outline these.

Use strongly typed scalars or a mix of both
-------------------------------------------

The simplest solution to the value-based promotion/casting issue would be to use
strongly typed Python scalars, i.e. Python floats are considered double precision
and Python integers are always considered the same as the default integer dtype.

This would be the simplest solution, however, it would lead to many upcasts when
working with arrays of ``float32`` or ``int16``, etc.  The solution for these cases

> would be to rely on in-place operations.

Reviewer 4: It may be good to write it as an explicit advantage of the proposed scheme that operations with a python scalar give the same result whether or not it is done in-place.

This NEP proposes to refactor the behaviour around two guiding principles:

1. The value must never influence the result type.
2. NumPy scalars or 0-D arrays must always lead to the same behaviour as
   their N-D counterparts.

We propose to remove all value-based logic and add special handling for
Python scalars to preserve some of the convenience that it provided.
This will mean that Python scalars are considered "weakly" typed.
A Python integer will always be converted to the NumPy dtype retaining the
behaviour that::

    np.array([1, 2, 3], dtype=np.uint8) + 1  # returns a uint8 array
    np.array([1, 2, 3], dtype=np.float32) + 2.  # returns a float32 array

but removing any dependence on the Python value itself.

The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
we expect that the behaviour in functions (promotion) will in practice be far
more important than the casting change itself.

Schema of the new proposed promotion rules
------------------------------------------

After the change, the promotions in NumPy will follow the schema below.
Promotion always occurs along the green lines:

> from left to right within their kind and to a higher kind only when

Author: You are right, will move the bubble.  It is super annoying, because technically, the bubble should be to the left of `uint64/int64`, but for practical reasons, numpy chooses to not put it there.
But... moving the bubble is right in practice for that reason.

NumPy value is integral, so we use at least ``float64``/``complex128``:

    np.uint16(3) + 3.0 == np.float64(6.0)
    np.int16(4) + 4j == np.complex128(4+4j)

But this does not happen for ``float`` to ``complex`` promotions, where
``float32`` and ``complex64`` have the same precision:

    np.float32(5) + 5j == np.complex64(5+5j)

The above table omits, ``bool``.  It is set below "integral", so that the
following hold::

    np.bool_(True) + 1 == np.int64(2)
    True + np.uint8(2) == np.uint8(3)

Motivation and Scope
====================

The motivation for changing the behaviour with respect to inspecting the value
of Python scalars and NumPy scalars/0-D arrays is, again, two-fold:

1. The special handling of NumPy scalars/0-D arrays as well as the value
   inspection can be very surprising to users.
2. The value-inspection logic is much harder to explain and implement.
   It is further harder to make it available to user defined DTypes through
   :ref:`NEP 42 <NEP42>`.
   Currently, this leads to a dual implementation of a new and an old (value
   sensitive) system.  Fixing this will greatly simplify the internal logic

>    and make results more consistent.

Author: Well, technically the current system is consistent between arrays and scalars.  It is inconsistent between 0-D and not 0-D, so not sure...

We believe that the proposal of "weak" Python scalars will help users by
providing a clearer mental model for which datatype an operation will
result in.
This model fits well with the preservation of array precisions that NumPy
currently often has, and also aggressively does for in-place operations::

    arr += value

Preserves precision so long "kind" boundaries are not crossed.

And while some users will probably miss the value inspecting behavior even for
those cases where it seems useful, it quickly leads to surprises.  This may be
expected::

    np.array([100], dtype=np.uint8) + 1000 == np.array([1100], dtype=np.uint16)

But the following will then be a surprise::

    np.array([100], dtype=np.uint8) + 200 == np.array([44], dtype=np.uint8)

> Considering this, we believe that the proposal follows the "principle of least

Reviewer 5: I don't understand this, as the table above shows that this behaviour is maintained?

Silent integer overflow is a viciously surreptitious source of bugs, and if I read the NEP correctly, we're increasing the surface for that by defaulting to lower integer types - e.g. `uint8(1) + 2`: `int64(3)` (old) -> `uint8(3)` (new)

Will return an int64 array because the type of ``np.int64(1)`` is strictly
honoured.

Impact on operators involving Python ``int``, ``float``, and ``complex``
------------------------------------------------------------------------

This NEP attempts to preserve the convenience that the old behaviour
gave when working with literal values.
The current value-based logic had some nice properties when "untyped",
literal Python scalars are involved::

  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
  np.array([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array

But led to complexity when it came to "unrepresentable" values:

  np.arange(10, dtype=np.int8) + 256  # returns int16
  np.array([1., 2.], dtype=np.float32) * 1e200  # returns float64

The proposal is to preserve this behaviour for the most part.  This is achieved
by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
in these operations.
Hoewver, to mitigate user surprises, we plan to make conversion to the new type
more strict:  This means that the results will be unchanged in the first
two examples.  For the second one, the results will be the following::

>   np.arange(10, dtype=np.int8) + 256  # raises a TypeError

Reviewer 5: Agreed with @rossbar. This is likely to break a ton of essentially harmless code. I think it would be great to raise on integer overflow (consistently, not just with python scalars), but I think it would be vastly preferable to only do that when we actually fall out of (u)int64.

It's laudable to be more precise on this (i.e. if someone _really_ wants to make sure their `int8` stays an `int8`), but I think it's going to be a very sharp edge for the large majority of users who do not care about the underlying representation of their ints, but will be hit with errors much more often.

The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
we expect that the behaviour in functions (promotion) will, in practice, be far
more important than the casting change itself.

Schema of the new proposed promotion rules
------------------------------------------

After the change, the promotions in NumPy will follow the schema below.
Promotion always occurs along the green lines:
from left to right within their kind and to a higher kind only when
necessary.
The result kind is always the largest kind of the inputs.
Note that NumPy considers ``float32`` to have a lower precision than ``int32``
and ``uint32``, but makes an exception for ``float64`` (and higher).

The Python scalars are inserted at the very left of each "kind" and the
Python integer does not distinguish signed and unsigned.
Note, that when the promoting a Python scalar with a dtype of lower kind
category (boolean, integral, inexact) with a higher one, we  use the
minimum/default precision: that is ``float64``, ``complex128`` or ``int64``
(``int32`` is used on some systems, e.g. windows).

.. figure:: _static/nep-0050-promotion-no-fonts.svg

>     :figclass: align-center

Reviewer 5: The horizontal lines in this graphic should have the same thickness and colour as the diagonal ones. At first glance it looks like `uint8` can only promote to `float16`, not `uint16`.

![grafik](https://user-images.githubusercontent.com/33685575/163495919-987c3816-0ce0-496e-abaa-e9c9263ea13f.png)

Author: Ah, got some alpha wrong or so, will fix, thanks.

potential changes in behavior in most relevant cases.

Impact on ``can_cast``
----------------------

Can cast will never inspect the value anymore.  So that the following results
are expected to change from ``True`` to ``False``::

  np.can_cast(100, np.uint8)
  np.can_cast(np.int64(100), np.uint8)
  np.can_cast(np.array(100, dtype=np.int64), np.uint8)

We expect that the impact of this change will be small compared to that of
the following changes.

.. note::

    The first example where the input is a Python scalar could be preserved
    to some degree, but this is not currently planned.

Impact on operators and functions involving NumPy arrays or scalars
-------------------------------------------------------------------

The main impact on operations not involving Python scalars (``float``, ``int``,
``complex``) will be that 0-D arrays and NumPy scalars will never behave
value-sensitive.
This removes currently surprising cases.  For example::

  np.arange(10, dtype=np.uint8) + np.int64(1)
  # and:
  np.add(np.arange(10, dtype=np.uint8), np.int64(1))

Will return an int64 array because the type of ``np.int64(1)`` is strictly

> honoured.

Reviewer 5: It's not clear from context ("currently surprising cases") if this "will" is describing current or future behaviour.

  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
  np.array([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array

But led to complexity when it came to "unrepresentable" values::

  np.arange(10, dtype=np.int8) + 256  # returns int16
  np.array([1., 2.], dtype=np.float32) * 1e200  # returns float64

The proposal is to preserve this behaviour for the most part.  This is achieved
by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
in these operations.
However, to mitigate user surprises, we plan to make conversion to the new type
more strict:  This means that the results will be unchanged in the first
two examples.  For the second one, the results will be the following::

  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
  np.array([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity

The second one will warn because ``np.float32(1e200)`` overflows to infinity.
It will then do the calculation with ``inf`` as normally.

.. admonition:: Behaviour in other libraries

   Overflowing in the conversion rather than raising an error is a choice;
   it is one that is the default in most C setups (similar to NumPy C can be

>    set up to raise an error due to the overflow, however).

Reviewer 5: I strongly [believe](https://github.com/numpy/numpy/issues/21219#issuecomment-1073184844) that we should never silently overflow, but promote to the end of the integer hierarchy before doing that (so if there's ever a portable and universally available `int128`, we should keep going IMO).

I realise that this is incompatible with the proposed move away from value-based promotions, but to me that seems a step in the wrong direction (not talking about the internals and their complexity, but from a UX POV).

   It is also for example the behaviour of ``pytorch`` 1.10.

Backward compatibility
======================

In general, code which only uses the default dtypes float64, or int32/int64
or more precise ones should not be affected.

However, the proposed changes will modify results in quite a few cases where
0-D or scalar values (with non-default dtypes) are mixed.
In many cases, these will be bug-fixes, however, there are certain changes
which may be particularly interesting.

The most important failure is probably the following example::

  arr = np.arange(100, dtype=np.uint8)  # storage array with low precision
  value = arr[10]

  # calculation continues with "value" without considering where it came from
  value * 100

Where previously the ``value * 100`` would cause an up-cast to int32/int64
(because value is a scalar).  The new behaviour will preserve the lower
precision unless explicitly dealt with (just as if ``value`` was an array).
This can lead to integer overflows and thus incorrect results beyond precision.
In many cases this may be silent, although NumPy usually gives warnings for the

> scalar operators.

Reviewer 5: This sound like a really bad idea to me. I'm sorry to be so negative on this, but this could well lead to silent and catastrophic failures & data corruption, and IMO poses a substantial risk of blemishing numpy's reputation if not handled with extreme care.

Author: Note that I am planning to add at least an *optional* warning tthat will notify of places where reduced precision occurs (I am not sure it would be 100%, but it should be good enough in practice).

Reviewer 6: Why will these warnings be optional?  This is the case I am by far the most worried about.

Author: Note that there are two warnings involved here:
* Overflow warnings are normally given for NumPy scalar operations (which are the main affected ones here).  (They are not given for array operations.)
* Warnings about the actual change, which I was thinking should probably be opt-in.

For the second, I was aiming for opt-in currently, because I think they will be very noisy and cumbersome to avoid.  E.g. for `float32 >= 2.1` where the behaviour might change, but practically never does.  Should we do the operation twice to check? Give a warning when it probably doesn't matter?

Maybe we could check whether always warnings should happen for more particularly surprising changes, right now I hoped the scalar integer overflow warnings cover the worst of it though.

Similarliy, if the storage array is float32 a calculation may retain the lower
float32 precision rather than use the default float64.

Further issues can occur.  For example:

* Floating point comparisons, especially equality, may change when mixing
  precisions::

     np.float32(1/3) == 1/3  # was False, will be True.

* Certain operations are expected to start failing::

     np.array([1], np.uint8) * 1000

>      np.array([1], np.uint8) == 1000  # possibly also

Reviewer 5: Why "possibly"? What does it depend on?

  to protect users in cases where previous value-based casting led to an
  upcast.  (Failures occur when converting ``1000`` to a ``uint8``.)

* Floating point overflow may occur in odder cases::

     np.float32(1e-30) * 1e50  # will return ``inf`` and a warning

  Because ``np.float32(1e50)`` returns ``inf``.  Previously, this would return
  a double precision result even if the ``1e50`` was not a 0-D array

In other cases, increased precision may occur.  For example::

  np.multiple(float32_arr, 2.)
  float32_arr * np.float64(2.)

> Will both return a float64 rather than float32.  This improves precision but

Reviewer 5: I thought python scalars would be treated as "weak"? Why does the first one upcast to float64?

Author: I think I use "Python scalars" for `int`, `float`, `complex`.  So the ones builtin with Python, not the *NumPy* scalars.  But maybe I can try to avoid that term to avoid the tricky distinction.

  to worry about whether they want to "remember" that an input was a Python
  scalar initially.  On the other hand, it would add a the need for slightly
  different (or additional) logic for Python operators.
  (Technically, probably as a flag to the ufunc dispatching mechanism to toggle
  the weak logic.)
* ``__array_ufunc__`` is often used on its own to provide Python operator
  support for array-likes implementing it.  If operators are special, these
  array-likes may need a mechanism to match NumPy (e.g. a kwarg to ufuncs to
  enable weak promotion.)

NumPy scalars could be special
------------------------------

Many users expect that NumPy scalars should be different from NumPy
arrays, in that ``np.uint8(3) + 3`` should return an ``int64`` (or Python
integer), when ``uint8_arr + 3`` preserves the ``uint8`` dtype.

This alternative would be very close to the current behaviour for NumPy scalars
but it would cement a distinction between arrays and scalars (NumPy arrays
are "stronger" than Python scalars, but NumPy scalars are not).

Such a distinction is very much possible, however, at this time NumPy will

> often (and silently) convert 0-D arrays to scalars.

Reviewer 5: To me, it's more important that 0-D arrays behave consistently to their N-D counterparts than that they behave consistently with scalars. Some libraries such as CVXPY have tightened this distinction in recent times, and I think this is a healthy choice.

A numpy 0-D array has a completely different API than a python scalar, and liberally casting one to the other is not beneficial in my opinion. I think they should be _compatible_, but that types should be maintained, e.g. I expect a function operating on scalar values to maintain the input type (0-D array -> 0-D array; python scalar -> python scalar; without silently changing between them).

Author: Yes, but NumPy does not currently have that choice, because we create scalars from 0-D arrays silently all the time.  If scalars were distinct objects, you only got with a special operation.  I might agree.

This NEP is only about the *dtype* of the array or scalar, not about this behaviour.  I agree that it should be cleaned up, but it is another set of worms.  But I also think that NumPy scalars have a `dtype` and because of that should behave the same as NumPy arrays for promotion.

2. Unlike JAX tensors, NumPy arrays are mutable, so assignment may need to
   cause it to be strongly typed?

A flag will likely be useful as an implementation detail (e.g. in ufuncs),
however, as of now we do not expect to have this as user API.
The main reason is that such a flag may be surprising for users if it is
passed out as a result from a function, rather than used only very localized.

.. admonition:: TODO

    Before accepting the NEP it may be good to discuss this issue further.
    Libraries may need clearer patterns to "propagate" the "weak" type, this
    could just be an ``np.asarray_or_literal()`` to preserve Python scalars,
    or a pattern of calling ``np.result_type()`` before ``np.asarray()``.

Keep using value-based logic for Python scalars
-----------------------------------------------

Some of the main issues with the current logic arise, because we apply it
to NumPy scalars and 0-D arrays, rather than the application to Python scalars.
We could thus consider to keep inspecting the value for Python scalars.

We reject this idea on the grounds that it will not remove the surprises
given earlier::

    np.uint8(100) + 1000 == np.uint16(1100)

>     np.uint8(100) + 200 == np.uint8(44)

Reviewer 5: I'd argue that the first one is fine and the second should be `np.uint16(300)` (more details in the comments above).

1. Values must never influence result type.
2. NumPy scalars and 0-D arrays should behave consistently with their
   N-D counterparts.

We propose to remove all value-based logic and add special handling for
Python scalars to preserve some convenient behaviors.
Python scalars will be considered "weakly" typed.
When a NumPy array/scalar is combined with a Python scalar, it will
be converted to the NumPy dtype, such that::

    np.array([1, 2, 3], dtype=np.uint8) + 1  # returns a uint8 array
    np.array([1, 2, 3], dtype=np.float32) + 2.  # returns a float32 array

There will be no dependence on the Python value itself.

The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
we expect that the behaviour in functions (promotion) will, in practice, be far
more important than the casting change itself.

Schema of the new proposed promotion rules
------------------------------------------

After the change, the promotions in NumPy will follow the schema below.
Promotion always occurs along the green lines:
from left to right within their kind and to a higher kind only when
necessary.
The result kind is always the largest kind of the inputs.

> Note that NumPy considers ``float32`` to have a lower precision than ``int32``

Author: The reason is that float32 cannot exactly represent all int32 values, so it makes sense.  The implication in the figure is that I moved float32 slightly to the left of int32.

It is not really safe to cast int32 to float32 or int64 to float64 and because of that promotion avoids int32 to float32 (it goes to float64).  But we make the exception that we consider int64 to float64 a safe cast and a valid promotion (even though it arguably loses precision).

> Note that NumPy considers ``float32`` to have a lower precision than ``int32``

Author: The implication on the figure is that the float32 bubble is to the left of the in32 to bubble

minimum/default precision: that is ``float64``, ``complex128`` or ``int64``
(``int32`` is used on some systems, e.g. windows).

.. figure:: _static/nep-0050-promotion-no-fonts.svg
    :figclass: align-center

See the next section for examples which clarify the proposed behavior.
Further examples with a comparison to the current behavior can be found
in the table below.

Examples of new behaviour
-------------------------

Since the schema and logic are difficult to read with respect to some cases,
these are examples of the new behaviour.  Below, the Python integer has no
influence on the result type::

    np.uint8(1) + 1 == np.uint8(2)
    np.int16(2) + 2 == np.int16(4)

In the following the Python ``float`` and ``complex`` are "inexact", but the
NumPy value is integral, so we use at least ``float64``/``complex128``::

    np.uint16(3) + 3.0 == np.float64(6.0)
    np.int16(4) + 4j == np.complex128(4+4j)

But this does not happen for ``float`` to ``complex`` promotions, where
``float32`` and ``complex64`` have the same precision::

    np.float32(5) + 5j == np.complex64(5+5j)

Note that the schema omits, ``bool``.  It is set below "integral", so that the
following hold::

>     np.bool_(True) + 1 == np.int64(2)

Author: Hmmm, certainly also plausible.  My gut feeling is that this is the right choice, because:
```
np.int8(3) + 1.
```
should be float64 I think, since there is no clear logic to prefer float32 (or float16?).

     - ``array([False])``
     - *unchanged*
   * - ``array([0.1], float32) == float64(0.1)``
     - ``array([ True])``
     - ``array([False])``  [T7]_
   * - ``array([1.], float32) + 3``
     - ``array([4.], float32)``
     - *unchanged*
   * - ``array([1.], float32) + int64(3)``
     - ``array([4.], float32)``
     - ``array([4.], float64)``  [T8]_

.. [T1] New behaviour honours the dtype of the ``uint8`` scalar.
.. [T2] Current NumPy ignores the precision of 0-D arrays or NumPy scalars
        when combined with arrays.
.. [T3] Current NumPy ignores the precision of 0-D arrays or NumPy scalars
        when combined with arrays.
.. [T4] Old behaviour uses ``uint16`` because ``300`` does not fit ``uint8``,
        new behaviour raises an error for the same reason.
.. [T5] ``300`` cannot be converted to ``uint8``.
.. [T6] ``np.float32(3e100)`` overflows to infinity.
.. [T7] ``0.1`` loses precision when cast to ``float32``, but old behaviour
        casts the ``float64(0.1)`` and then matches.
.. [T8] NumPy promotes ``float32`` and ``int64`` to ``float64``.  The old
        behaviour ignored the ``int64`` here.

Motivation and Scope
====================

> The motivation for changing the behaviour with respect to inspecting the value

Reviewer 6: I see this is mentioned below; perhaps a forward reference to that section.

> The motivation for changing the behaviour with respect to inspecting the value

Author: Adding a third item in the "motivation", that this will align us with other projects (and then link to there)

   sensitive) system.  Fixing this will greatly simplify the internal logic
   and make results more consistent.

We believe that the proposal of "weak" Python scalars will help users by
providing a clear mental model for which datatype an operation will
result in.
This model fits well with the preservation of array precisions that NumPy
currently often follows, and also uses for in-place operations::

    arr += value

Preserves precision as long as "kind" boundaries are not crossed (otherwise
an error is raised).

While some users will potentially miss the value inspecting behavior even for
those cases where it seems useful, it quickly leads to surprises.  This may be
expected::

    np.array([100], dtype=np.uint8) + 1000 == np.array([1100], dtype=np.uint16)

But the following will then be a surprise::

    np.array([100], dtype=np.uint8) + 200 == np.array([44], dtype=np.uint8)

Considering that the proposal aligns with the behavior of in-place operands
and avoids the surprising switch in behavior that only sometimes avoids
overflow in the result,
we believe that the proposal follows the "principle of least surprise".

Usage and Impact
================

> This NEP is expected to be implemented with **no** transition period that warns

Reviewer 6: Yikes; this has the potential to break some codes out there, without any obvious way to fix them.  We should probably, at least, make a major version release.

detailed in the backwards compatibility section.

The solution to this will be an *optional* warning mode capable of notifying
users of potential changes in behavior.
This mode is expected to generate many harmless warnings, but provide a way
to systematically vet code and track down changes if problems are observed.

Impact on ``can_cast``
----------------------

`can_cast` will never inspect the value anymore.  So that the following results
are expected to change from ``True`` to ``False``::

  np.can_cast(np.int64(100), np.uint8)
  np.can_cast(np.array(100, dtype=np.int64), np.uint8)
  np.can_cast(100, np.uint8)

We expect that the impact of this change will be small compared to that of
the following changes.

.. note::

    The last example where the input is a Python scalar _may_ be preserved
    since ``100`` can be represented by a ``uint8``.

Impact on operators and functions involving NumPy arrays or scalars
-------------------------------------------------------------------

The main impact on operations not involving Python scalars (``float``, ``int``,
``complex``) will be that 0-D arrays and NumPy scalars will never behave

> value-sensitive.

Reviewer 6: ```suggestion
value-sensitively.
```

Although, this phrasing is slightly odd; can reword perhaps to something like "that operations on 0-D arrays and NumPy scalars will never depend on their values"

This removes currently surprising cases.  For example::

  np.arange(10, dtype=np.uint8) + np.int64(1)
  # and:
  np.add(np.arange(10, dtype=np.uint8), np.int64(1))

Will return an ``int64`` array in the future because the type of
``np.int64(1)`` is strictly honoured.
Currently a ``uint8`` array is returned.

Impact on operators involving Python ``int``, ``float``, and ``complex``
------------------------------------------------------------------------

> This NEP attempts to preserve the convenience that the old behaviour
> gave when working with literal values.

Reviewer 6: ```suggestion
This NEP attempts to preserve the convenience of the old behaviour
when working with literal values.
```

The current value-based logic had some nice properties when "untyped",
literal Python scalars are involved::

  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
  np.array([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array

But led to surprises when it came to "unrepresentable" values::

  np.arange(10, dtype=np.int8) + 256  # returns int16
  np.array([1., 2.], dtype=np.float32) * 1e200  # returns float64

The proposal is to preserve this behaviour for the most part.  This is achieved
by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
in operations.

> However, to avoid user surprises, we plan to make conversion to the new type

Reviewer 6: ```suggestion
However, to avoid surprises, we plan to make conversion to the new type
```

more strict:  The results will be unchanged in the first two examples,
but in the second one, it will change the following way::

  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
  np.array([1., 2.], dtype=np.float32) * 1e200  # warning and returns infinity

The second one warns because ``np.float32(1e200)`` overflows to infinity.

> It will then do the calculation with ``inf`` as normally.

Reviewer 6: ```suggestion
It will then continue to do the calculation with ``inf`` as usual.
```
